ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -0,0 +1,185 @@
+
+	if rayJobInstance, err = r.getRayJobInstance(ctx, request); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	var rayClusterInstance *rayv1alpha1.RayCluster
+	if rayClusterInstance, err = r.getOrCreateRayClusterInstance(ctx, rayJobInstance); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log.Info("Updated RayCluster")
+	rayJobInstance.Status.RayClusterStatus = rayClusterInstance.Status
+
+	var clientURL string
+	if clientURL, err = common.FetchDashboardURL(ctx, r.Client, rayClusterInstance); err != nil || clientURL == "" {
+		if clientURL == "" {
+			err = fmt.Errorf("Empty dashboardURL")
+		}
+		return ctrl.Result{}, err
+	}
+
+	rayDashboardClient := utils.RayDashboardClient{}
+	rayDashboardClient.InitClient(clientURL)
+
+	if rayJobInstance.Status.JobId == "" {
+		// Submit the job if no id set
+		jobId, err := rayDashboardClient.SubmitJob(rayJobInstance)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		log.Info("Job succesfully submmited", "jobId", jobId)
+		rayJobInstance.Status.JobId = jobId
+		rayJobInstance.Status.JobStatus = rayv1alpha1.JobStatus_Pending
+		if err = r.Status().Update(ctx, rayJobInstance); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, nil
+	}
+
+	// If jobId set already, just check job status
+	jobInfo, err := rayDashboardClient.GetJobInfo(rayJobInstance.Status.JobId)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	rayJobInstance.Status.JobStatus = jobInfo.JobStatus
+	if err = r.Status().Update(ctx, rayJobInstance); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *RayJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&rayv1alpha1.RayJob{}).
+		Complete(r)
+}
+
+func (r *RayJobReconciler) getRayJobInstance(ctx context.Context, request ctrl.Request) (*rayv1alpha1.RayJob, error) {
+	rayJobInstance := &rayv1alpha1.RayJob{}
+	if err := r.Get(ctx, request.NamespacedName, rayJobInstance); err != nil {
+		if errors.IsNotFound(err) {
+			rayServiceLog.Info("Read request instance not found error!")
+		} else {
+			rayServiceLog.Error(err, "Read request instance error!")
+		}
+		// Error reading the object - requeue the request.
+		return nil, err
+	}
+	return rayJobInstance, nil
+}
+
+// TODO: select existing rayclusters by ClusterSelector
+func (r *RayJobReconciler) getOrCreateRayClusterInstance(ctx context.Context, rayJobInstance *rayv1alpha1.RayJob) (*rayv1alpha1.RayCluster, error) {
+	// Update ray cluster
+	var rayClusterInstanceName string
+	if rayJobInstance.Status.RayClusterName != "" {
+		rayClusterInstanceName = rayJobInstance.Status.RayClusterName
+	} else {
+		rayClusterInstanceName = fmt.Sprintf("%s%s%s", rayJobInstance.Name, rayClusterSuffix, rand.String(5))
```

Comment:
here we also need to consider if the status update fails, do we create multiple clusters.

Reply:
I agree with you. I saw you use two status (pending/active cluster) to record. I may submit another PR to resolve this issue